### DISCUSSION TITLE
/simplify-communication-visual-studio-code-extension-webview/

### DISCUSSION THREAD

**Discussion by comment-migrator** [estruyf/eliostruyf-comments#604] [Comments]
Simplify Visual Studio Code extension webview communication
        
[https://www.eliostruyf.com/simplify-communication-visual-studio-code-extension-webview/](https://www.eliostruyf.com/simplify-communication-visual-studio-code-extension-webview/)

**Comment by Askhento** [comment]
Thanks for a great idea.  I made a similar, but very crude version which waits for the first response.
What about communication between webviews?

**Reply by estruyf** [reply]
I haven't done anything to make it work between webviews, although the process to make it work would be more like a guideline/blueprint.

If webview 1 requested something from webview 2, the flow would go as follows:

1. Webview 1 does the request by providing a request ID to async/await it
2. Listener on the extension host picks it up and forwards it to webview 2
3. Webview 2 listens to post messages, and when it picks up the message, it handles it
4. Webview 2 sends back a post message to the listener
5. Listener picks up the message, and does a post message with the request ID which was originally created in step 1

So the request logic to async/await would work between the webview and its listener, but when there is another webview involved, I'm afraid that would be a bit complicated, because of the decoupled async logic.

**Reply by Askhento** [reply]
Hey thanks for the response, that was really helpful! Finally request could be awaited!
I've implemented this behaviour with redirect inside MessageHandler class.

Here is an interface I use: 
```typescript
export interface MessageHandlerData {
    requestId?: string;
    target: string;
    origin?: string;
    command: string;
    payload: any;
}
```
If target !== origin this mean we are redirecting. I always include origin in requests so that after redirect i know where to send it back. 
Also as a target I have 'all' so that it casts messages to everyone extension host included. All webviews use their own MessageHandler implementation and they also can deal with requests from extension (i use it to retrieve logs). Would love to share, but my code style is a garbage((

**Reply by estruyf** [reply]
Hi @Askhento, feel free to share the code. BTW, in which extension are you using it?
